fdb-record-layer-core/src/main/proto/record_query_plan.proto

```diff
@@ -0,0 +1,1352 @@
+    PArithmeticValue arithmetic_value = 3;
+    PConditionSelectorValue condition_selector_value = 4;
+    PConstantObjectValue constant_object_value = 5;
+    PConstantValue constant_value = 6;
+    PCountValue count_value = 7;
+    PDerivedValue derived_value = 8;
+    PEmptyValue empty_value = 9;
+    PExistsValue exists_value = 10;
+    PFieldValue field_value = 11;
+    PIndexedValue indexed_value = 12;
+    PMaxEverLongValue max_ever_long_value = 13;
+    PMinEverLongValue min_ever_long_value = 14;
+    PInOpValue in_op_value = 15;
+    PLikeOperatorValue like_operator_value = 16;
+    PLiteralValue literal_value = 17;
+    PNotValue not_value = 18;
+    PNullValue null_value = 19;
+    PNumericAggregationValue.PSum numeric_aggregation_value_sum = 20;
+    PNumericAggregationValue.PAvg numeric_aggregation_value_avg = 21;
+    PNumericAggregationValue.PMin numeric_aggregation_value_min = 22;
+    PNumericAggregationValue.PMax numeric_aggregation_value_max = 23;
+    PObjectValue object_value = 24;
+    POfTypeValue of_type_value = 25;
+    PPatternForLikeValue pattern_for_like_value = 26;
+    PPickValue pick_value = 27;
+    PPromoteValue promote_value = 28;
+    PQuantifiedObjectValue quantified_object_value = 29;
+    PQueriedValue queried_value = 30;
+    PRankValue rank_value = 31;
+    PRecordConstructorValue record_constructor_value = 32;
+    PRecordTypeValue record_type_value = 33;
+    PBinaryRelOpValue binary_rel_op_value = 34;
+    PUnaryRelOpValue unary_rel_op_value = 35;
+    PVariadicFunctionValue variadic_function_value = 36;
+    PVersionValue version_value = 37;
+  }
+}
+
+message PAbstractArrayConstructorValue {
+  repeated PValue children = 1;
+  optional PType element_type = 2;
+}
+
+message PLightArrayConstructorValue {
+  optional PAbstractArrayConstructorValue super = 1;
+}
+
+message PAndOrValue {
+  enum POperator {
+    AND = 1;
+    OR = 2;
+  }
+  optional string function_name = 1;
+  optional PValue left_child = 2;
+  optional PValue right_child = 3;
+  optional POperator operator = 4;
+}
+
+message PArithmeticValue {
+  enum PPhysicalOperator {
+      ADD_II = 1;
+      ADD_IL = 2;
+      ADD_IF = 3;
+      ADD_ID = 4;
+      ADD_IS = 5;
+      ADD_LI = 6;
+      ADD_LL = 7;
+      ADD_LF = 8;
+      ADD_LD = 9;
+      ADD_LS = 10;
+      ADD_FI = 11;
+      ADD_FL = 12;
+      ADD_FF = 13;
+      ADD_FD = 14;
+      ADD_FS = 15;
+      ADD_DI = 16;
+      ADD_DL = 17;
+      ADD_DF = 18;
+      ADD_DD = 19;
+      ADD_DS = 20;
+      ADD_SI = 21;
+      ADD_SL = 22;
+      ADD_SF = 23;
+      ADD_SD = 24;
+      ADD_SS = 25;
+
+      SUB_II = 26;
+      SUB_IL = 27;
+      SUB_IF = 28;
+      SUB_ID = 29;
+      SUB_LI = 30;
+      SUB_LL = 31;
+      SUB_LF = 32;
+      SUB_LD = 33;
+      SUB_FI = 34;
+      SUB_FL = 35;
+      SUB_FF = 36;
+      SUB_FD = 37;
+      SUB_DI = 38;
+      SUB_DL = 39;
+      SUB_DF = 40;
+      SUB_DD = 41;
+
+      MUL_II = 42;  // my favorite
```

Comment:
Your comment is actually geared towards how that `Value` is implemented. You can encapsulate the input arguments to bring them to the max type and then perform the operation or you can have a physical operator for every possible combination. We went with the latter.

This PR does not attempt to be optimal here, we try to just reflect the implementation in proto: If the java class follows a particular structure, the proto structure will attempt to mimic that as close as possible.

Reply:
I think what this PR shows is that the approach of enumerating out the various types quickly falls to the tyranny of the fundamental counting principle. I've found this pretty annoying when trying to update Java code (if I'm perfectly honest), but I think this is worse because it it bakes this choice into the serialization format which makes it harder to change in the future. If we, for example, want to introduce a new "type" here for the left and right arguments, then we'll have to introduce them without changing the enum numbers of any of the existing types.

I think this is a place where it would be better for the serialization to diverge from the POJO structure (or for the POJO structure to similarly change)

Reply:
I think I found a compromise which removes the hard-to-maintain switch statements by keeping some enum maps.